lib/fizzy/utf8.cpp

```diff
@@ -0,0 +1,159 @@
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+100000..U+10FFFF | F4         | 80..8F      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ */
+
+// Ranges for 2nd byte
+enum class Rule
+{
+    Range80BF,  // 80..BF
+    RangeA0BF,  // A0..BF
+    Range809F,  // 80..9F
+    Range90BF,  // 90..BF
+    Range808F,  // 80..8F
+};
+
+namespace fizzy
+{
+bool utf8_validate(const uint8_t* pos, const uint8_t* end) noexcept
+{
+    while (pos < end)
+    {
+        unsigned required_bytes = 1;
+        auto rule = Rule::Range80BF;
+
+        const uint8_t byte1 = *pos++;
+        if (byte1 <= 0x7F)
+            // Shortcut for valid ASCII (also valid UTF-8)
+            continue;
+        else if (byte1 < 0xC2)
+            // NOLINTNEXTLINE(bugprone-branch-clone)
+            return false;
+        else if (byte1 <= 0xDF)
+        {
+            required_bytes = 2;
+            rule = Rule::Range80BF;
+        }
+        else if (byte1 == 0xE0)
+        {
+            required_bytes = 3;
+            rule = Rule::RangeA0BF;
+        }
+        else if (byte1 <= 0xEC)
+        // NOLINTNEXTLINE(bugprone-branch-clone)
+        {
+            required_bytes = 3;
+            rule = Rule::Range80BF;
+        }
+        else if (byte1 == 0xED)
+        {
+            required_bytes = 3;
+            rule = Rule::Range809F;
+        }
+        else if (byte1 <= 0xEF)
+        {
+            required_bytes = 3;
+            rule = Rule::Range80BF;
+        }
+        else if (byte1 == 0xF0)
+        {
+            required_bytes = 4;
+            rule = Rule::Range90BF;
+        }
+        else if (byte1 <= 0xF3)
+        {
+            required_bytes = 4;
+            rule = Rule::Range80BF;
+        }
+        else if (byte1 == 0xF4)
+        {
+            required_bytes = 4;
+            rule = Rule::Range808F;
+        }
+        else
+            return false;
+
+        // At this point need to read at least one more byte
+        if ((pos + required_bytes - 1) > end)
+            return false;
+
+        assert(required_bytes > 1);
+
+        // Byte2 may have exceptional encodings
+        const uint8_t byte2 = *pos++;
+        switch (rule)
+        {
+        case Rule::Range80BF:
+            if (byte2 < 0x80 || byte2 > 0xBF)
+                return false;
+            break;
+        case Rule::RangeA0BF:
+            if (byte2 < 0xA0 || byte2 > 0xBF)
+                return false;
+            break;
+        case Rule::Range809F:
+            if (byte2 < 0x80 || byte2 > 0x9F)
+                return false;
+            break;
+        case Rule::Range90BF:
+            if (byte2 < 0x90 || byte2 > 0xBF)
+                return false;
+            break;
+        case Rule::Range808F:
+            if (byte2 < 0x80 || byte2 > 0x8F)
+                return false;
+            break;
+        default:
```

Comment:
Default case not needed if using `enum class` locally.

Reply:
Fixed.